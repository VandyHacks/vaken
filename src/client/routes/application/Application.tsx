import React, { useContext, FunctionComponent, useState, useEffect, useCallback, FC } from 'react';
import styled from 'styled-components';
import { useImmer } from 'use-immer';
import config from '../../assets/application';
import { Collapsible } from '../../components/Containers/Collapsible';
import { ActionButtonContext } from '../../contexts/ActionButtonContext';
import { HeaderButton } from '../../components/Buttons/HeaderButton';
import { InputProps } from '../../components/Input/TextInput';
import { useUpdateMyApplicationMutation, useMyApplicationQuery } from '../../generated/graphql';
import { GraphQLErrorMessage } from '../../components/Text/ErrorMessage';

export interface ConfigSection {
	category: string;
	fields: ConfigField[];
	title: string;
}

export interface ConfigField {
	Component: FC<InputProps>;
	default?: string;
	fieldName: string;
	note?: string;
	optional?: boolean;
	options?: string[];
	other?: boolean;
	placeholder?: string;
	prompt?: string;
	required?: boolean;
	title: string;
	validation?: string;
}

export const StyledForm = styled.form`
	display: flex;
	flex-flow: column nowrap;

	fieldset {
		margin-top: 0.4rem;
	}
`;

export const StyledQuestion = styled.label`
	display: flex;
	flex-flow: column nowrap;
	font-size: 1rem;

	& > input {
		border-radius: 6px;
		background: white;
	}
`;

export const StyledQuestionPadContainer = styled.div`
	margin-bottom: 0.4rem;
`;

export const FieldPrompt = styled.h3`
	font-weight: lighter;
	font-size: 1em;
`;

export const FieldNote = styled.span`
	font-style: italic;
	font-weight: lighter;
	font-size: 1em;
`;

export const FieldTitle = styled.span`
	display: inline-block;
	font-style: normal;
	font-size: 1em;
	line-height: 140%;
`;

export const Application: FunctionComponent<{}> = (): JSX.Element => {
	const { update: setActionButton } = useContext(ActionButtonContext);
	const [openSection, setOpenSection] = useState('');
	const [input, setInput] = useImmer<{ answer: string; question: string }[]>([]);
	const updateApplication = useUpdateMyApplicationMutation();
	const { data, error, loading } = useMyApplicationQuery();

	// [{ObjectID: firstName, input: 'hello', changed: false}]
	// Only update changed QuestionIDs
	const createOnChangeHandler = (fieldName: string): ((value: string) => void) => value => {
		void setInput(draft => {
			const element = draft.find(el => el.question === fieldName);
			// FIXME: add a changed field to the question
			if (!element) {
				draft.push({ answer: value, question: fieldName });
			} else {
				element.answer = value;
			}
		});
	};

	const valueHandler = (fieldName: string): string => {
		const element = input.find(el => el.question === fieldName);
		return element ? element.answer : '';
	};

	useEffect((): (() => void) => {
		if (setActionButton)
			setActionButton(
				<HeaderButton
					onClick={() => {
						updateApplication({ variables: { input } });
					}}
					width="8em">
					Submit
				</HeaderButton>
			);

		return () => {
			if (setActionButton) setActionButton(undefined);
		};
	}, [input, setActionButton, updateApplication]);

	useEffect((): void => {
		if (!loading && data && data.me) {
			const { application } = data.me;
			setInput(draft => {
				application.forEach(({ answer, question }) => {
					const element = draft.find(el => el.question === question);
					if (element) {
						element.answer = answer;
					} else {
						draft.push({ answer, question });
					}
				});
			});
		}
	}, [data, loading, setInput]);

	const toggleOpen = useCallback(
		(e: React.MouseEvent<HTMLButtonElement>): void => {
			const { id } = e.currentTarget;
			setOpenSection(openSection === id ? '' : id);
			e.preventDefault();
		},
		[openSection]
	);

	// if error getting application input
	if (error) {
		console.log(error);
		return <GraphQLErrorMessage text={JSON.stringify(error)} />;
	}

	return (
		<StyledForm>
			{config.map(({ fields, title }: ConfigSection) => (
				<Collapsible onClick={toggleOpen} open={openSection === title} title={title} key={title}>
					{fields.map(field => (
						<StyledQuestion key={field.title} htmlFor={field.title}>
							<StyledQuestionPadContainer>
								{field.title}
								{field.note ? <FieldNote>{` - ${field.note}`}</FieldNote> : null}
							</StyledQuestionPadContainer>
							{field.prompt ? <FieldPrompt>{field.prompt}</FieldPrompt> : null}
							<field.Component
								setState={createOnChangeHandler(field.title)}
								value={valueHandler(field.title)}
								{...field}
								id={field.title}
							/>
						</StyledQuestion>
					))}
				</Collapsible>
			))}
		</StyledForm>
	);
};

export default Application;
