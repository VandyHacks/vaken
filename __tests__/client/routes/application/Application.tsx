import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import Application from '../../../../src/client/routes/application/Application';

it('Application initializes', () => {
	expect(Application).toBeTruthy();
	/*
	const component = renderer.create(<Application />).toJSON();

    expect(component).toMatchSnapshot();
    */
});
