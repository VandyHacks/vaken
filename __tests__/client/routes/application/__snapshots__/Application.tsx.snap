// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Application  renders correctly 1`] = `
<form
  className="sc-ckVGcZ bGOBYD"
>
  <div
    className="sc-jTzLTM cWhdTY"
  >
    <button
      className="sc-VigVT eIBicm"
      id="Demographic"
      onClick={[Function]}
    >
      Demographic
      <img
        alt="arrow"
        src={[Function]}
      />
    </button>
    <div
      className="sc-fjdhpX lgHlcu"
    >
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="First Name"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          First Name
        </div>
        <input
          className="sc-bdVaJa fREsov"
          id="First Name"
          onChange={[Function]}
          placeholder="John"
          value=""
        />
      </label>
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="Last Name"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          Last Name
        </div>
        <input
          className="sc-bdVaJa fREsov"
          id="Last Name"
          onChange={[Function]}
          placeholder="Smith"
          value=""
        />
      </label>
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="Date of Birth"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          Date of Birth
        </div>
        <input
          className="sc-bxivhb jiVfTZ"
          id="Date of Birth"
          onChange={[Function]}
          placeholder="01/01/2000"
          required={false}
          type="date"
          value=""
        />
      </label>
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="Graduation Year"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          Graduation Year
        </div>
        <fieldset>
          <div
            className="sc-htpNat ghQiwQ"
          >
            <input
              checked={false}
              id={2019}
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor={2019}
            >
              2019
            </label>
            <input
              checked={false}
              id={2020}
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor={2020}
            >
              2020
            </label>
            <input
              checked={false}
              id={2021}
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor={2021}
            >
              2021
            </label>
            <input
              checked={false}
              id={2022}
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor={2022}
            >
              2022
            </label>
          </div>
        </fieldset>
      </label>
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="Ethnicity"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          Ethnicity
        </div>
        <h3
          className="sc-hMqMXs cEhjZi"
        >
          Are you Spanish, Hispanic, or Latino?
        </h3>
        <fieldset>
          <div
            className="sc-htpNat ghQiwQ"
          >
            <input
              checked={false}
              id="Yes"
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor="Yes"
            >
              Yes
            </label>
            <input
              checked={true}
              id="No"
              onChange={[Function]}
              type="radio"
            />
            <label
              htmlFor="No"
            >
              No
            </label>
          </div>
        </fieldset>
      </label>
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="Race"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          Race
          <span
            className="sc-kEYyzF iVdoMM"
          >
             - Choose all that apply.
          </span>
        </div>
        <fieldset>
          <div
            className="sc-bwzfXH eWyhgy"
          >
            <div>
              <input
                checked={false}
                id="American Indian or Alaskan Native"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="American Indian or Alaskan Native"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                American Indian or Alaskan Native
              </label>
            </div>
            <div>
              <input
                checked={false}
                id="Asian"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="Asian"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                Asian
              </label>
            </div>
            <div>
              <input
                checked={false}
                id="Black or African American"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="Black or African American"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                Black or African American
              </label>
            </div>
            <div>
              <input
                checked={false}
                id="White or Caucasian"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="White or Caucasian"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                White or Caucasian
              </label>
            </div>
            <div>
              <input
                checked={false}
                id="Native Hawaiian or Other Pacific Islander"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="Native Hawaiian or Other Pacific Islander"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                Native Hawaiian or Other Pacific Islander
              </label>
            </div>
            <div>
              <input
                checked={false}
                id="Other"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="Other"
              >
                <icon-mock
                  height={24}
                  width={24}
                />
                Other
              </label>
            </div>
          </div>
        </fieldset>
      </label>
    </div>
  </div>
  <div
    className="sc-jTzLTM cWhdTY"
  >
    <button
      className="sc-VigVT eIBicm"
      id="Education"
      onClick={[Function]}
    >
      Education
      <img
        alt="arrow"
        src={[Function]}
      />
    </button>
    <div
      className="sc-fjdhpX lgHlcu"
    >
      <label
        className="sc-jKJlTe iibufK"
        htmlFor="School"
      >
        <div
          className="sc-eNQAEJ gMaoCC"
        >
          School
        </div>
        <input
          className="sc-bdVaJa fREsov"
          id="School"
          list="Vanderbilt University"
          onChange={[Function]}
          placeholder="If not found, please type the name."
          type="text"
          value=""
        />
        <datalist
          id="Vanderbilt University"
        >
          <option
            value="Vanderbilt University"
          />
          <option
            value=""
          />
        </datalist>
      </label>
    </div>
  </div>
</form>
`;
